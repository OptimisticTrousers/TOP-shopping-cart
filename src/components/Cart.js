import CartItem from "./CartItem"
import logo from '../assets/optimistictrousers.jpg'

export default function Cart() {

    return (
        <div className="cart-container">
            <h1 className="cart-title">Your Cart</h1>
            <div className="cart">
                <CartItem image={logo}/>
                <CartItem image={logo}/>
                <CartItem image={logo}/>
                <CartItem image={logo}/>
                <CartItem image={logo}/>

            </div>
            <p>SUBTOTAL</p>
            <div className="buttons">
                <button type="button">Check Out</button>
                <button type="button">Back To Products</button>
            </div>
        </div>
    )
}