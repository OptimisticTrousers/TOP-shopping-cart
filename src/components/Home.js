import Hero from "./Hero"
import About from "./About"

export default function Home() {
    return (
        <div className="home">
            <Hero /> 
            <About />
        </div>
    )
}