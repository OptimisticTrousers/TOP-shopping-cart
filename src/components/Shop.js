export default function Shop() {
    return (
        <main>
            <h1>Get Clothing!</h1>
            <p>Times are tough. Liven up with our top-notch clothing!</p>
            <div className="products">

            </div>
        </main>
    )
}