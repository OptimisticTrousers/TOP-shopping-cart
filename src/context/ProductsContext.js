import React, { useContext, useState } from "react";

const ProductsContext = React.createContext()

export function useProducts() {
    const [products] = useContext(ProductsContext)

    return products
}

export function useProductAdder() {
    const [products] = useContext(ProductsContext)

    return products
}

export function ProductsProvider({children}) {

    const [products, setProducts] = useState([])

    function addProduct(){
        
    }

    return (
        <ProductsContext.Provider value={{products, addProduct}}>
            {children}
        </ProductsContext.Provider>
    )
}