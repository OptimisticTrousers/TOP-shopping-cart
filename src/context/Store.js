import React, { useContext, useState } from "react"

const CartQuantityContext = React.createContext()
const CartContext = React.createContext()


export function useQuantity() {
    const {cartQuantity} = useContext(CartQuantityContext)

    return cartQuantity;
}

export function useAddToCart() {
    const {addToCart} = useContext(CartContext);

    return addToCart;
}

export function useCart() {
    const {cart} = useContext(CartContext)

    return cart
}

export function CartProvider({children}) {

    const [cartQuantity, setCartQuantity] = useState(0)
    const [cart, setCart] = useState([])


    const addToCart = (details) => {
        setCartQuantity(prevQuantity => prevQuantity + 1)
        setCart(prevCart => {

            const productIndex = prevCart.findIndex((product) => product.id === details.id)

            if(productIndex !== -1){
                const newCart = [...prevCart]

                newCart.splice(productIndex, 1, {...details, quantity: prevCart[productIndex].quantity + 1})

                return newCart;
            }
            return [...prevCart, {...details, quantity: 1}];
        })
    }

    return (
        <CartContext.Provider value={{cart, addToCart}}>
            <CartQuantityContext.Provider value={{cartQuantity}}>
                {children}
            </CartQuantityContext.Provider>
        </CartContext.Provider>
    )
}
