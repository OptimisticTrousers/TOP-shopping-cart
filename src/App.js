import React from "react";
import { Outlet } from "react-router-dom";
import Navbar from "./components/Navbar";
import Footer from "./components/Footer";
import { CartProvider } from "./context/Store";

export default function App() {

  return (
    <div className="container">
      <CartProvider>
        <Navbar />
        <main>
          <Outlet />
          <Footer />
        </main>
      </CartProvider>
    </div>
  );
}