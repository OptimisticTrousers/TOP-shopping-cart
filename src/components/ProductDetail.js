import { Link, useParams } from "react-router-dom"
import { useEffect, useState } from "react"
import leftArrow from '../assets/left-arrow.svg'
import { useAddToCart } from "../context/Store"

export default function ProductDetail(){

    const addToCart = useAddToCart()

    const params = useParams()

    const [details, setDetails] = useState({})

    useEffect(() => {
        fetch(`https://fakestoreapi.com/products/${params.id}`)
            .then(res => res.json())
            .then(data => setDetails(data))
    }, [])

    return (
        <div className="detail-container">
            <div className="image-detail-view">
                <img className="image-detail" src={details.image} alt={details.title}/>
            </div>
            <div className="product-details">
                <Link to="/shop">
                    <button type="button"><img className="left-arrow" src={leftArrow} alt="left arrow" /> Back To Products</button>
                </Link>
                <h1 className="product-title">{details.title}</h1>
                <p className="product-price">${details.price}</p>
                <p className="category-detail">Category: {details.category}</p>
                <p className="description-detail">{details.description}</p>
                <button onClick={() => addToCart({...details, quantity: 0})} className="add-to-cart-button" type="button">Add to Cart</button>
            </div>
        </div>
    )
}