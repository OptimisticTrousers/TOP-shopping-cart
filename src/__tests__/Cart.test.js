import React from 'react'
import {cleanup, fireEvent, render, screen, waitFor, waitForDomChange, waitForElementToBeRemoved} from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import Cart from '../components/Cart'
import ProductDetail from '../components/ProductDetail'
import {BrowserRouter, Routes, Route} from 'react-router-dom'
import * as Store from '../context/Store'
import App from '../App'
import Home from '../components/Home'
import Shop from '../components/Shop'


describe("CartItem", () => {

        beforeEach(() => {

        fetch.mockClear()

        fetch.mockResponseOnce(JSON.stringify([{
                    id: 1,
                    title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
                    price: 109.95,
                    description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                    rating: {
                    rate: 3.9,
                    count: 120
                    }
                },
                {
                    id: 2,
                    title: "Mens Casual Premium Slim Fit T-Shirts ",
                    price: 22.3,
                    description: "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
                    rating: {
                    rate: 4.1,
                    count: 259
                    }
                },
                {
                    id: 3,
                    title: "Mens Cotton Jacket",
                    price: 55.99,
                    description: "great outerwear jackets for Spring/Autumn/Winter, suitable for many occasions, such as working, hiking, camping, mountain/rock climbing, cycling, traveling or other outdoors. Good gift choice for you or your family member. A warm hearted love to Father, husband or son in this thanksgiving or Christmas Day.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg",
                    rating: {
                    rate: 4.7,
                    count: 500
                    }
                }]))
    })

    it("snapshot", () => {

        const {asFragment} = render(
            <BrowserRouter>
                <Store.CartProvider>
                    <Cart />
                </Store.CartProvider>
            </BrowserRouter>
        )

        expect(asFragment()).toMatchSnapshot()
    })

    it("rendering 3 cart items in cart", () => {

        Store.useCart = jest.fn(() => [{
                    id: 1,
                    title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
                    price: 109.95,
                    description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 3.9,
                    count: 120
                    }
                },
                {
                    id: 2,
                    title: "Mens Casual Premium Slim Fit T-Shirts ",
                    price: 22.3,
                    description: "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 4.1,
                    count: 259
                    }
                },
                {
                    id: 3,
                    title: "Mens Cotton Jacket",
                    price: 55.99,
                    description: "great outerwear jackets for Spring/Autumn/Winter, suitable for many occasions, such as working, hiking, camping, mountain/rock climbing, cycling, traveling or other outdoors. Good gift choice for you or your family member. A warm hearted love to Father, husband or son in this thanksgiving or Christmas Day.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 4.7,
                    count: 500
                    }
                }])

        render(
            <BrowserRouter>
                <Store.CartProvider>
                        <Cart />
                </Store.CartProvider>
            </BrowserRouter>
        )

        const cartItems = screen.queryAllByTestId('product-quantity')

        expect(cartItems.length).toBe(3)
    })

    it("correctly calculating subtotal", () => {

        Store.useCart = jest.fn(() => [{
                    id: 1,
                    title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
                    price: 109.95,
                    description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 3.9,
                    count: 120
                    }
                },
                {
                    id: 2,
                    title: "Mens Casual Premium Slim Fit T-Shirts ",
                    price: 22.3,
                    description: "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 4.1,
                    count: 259
                    }
                },
                {
                    id: 3,
                    title: "Mens Cotton Jacket",
                    price: 55.99,
                    description: "great outerwear jackets for Spring/Autumn/Winter, suitable for many occasions, such as working, hiking, camping, mountain/rock climbing, cycling, traveling or other outdoors. Good gift choice for you or your family member. A warm hearted love to Father, husband or son in this thanksgiving or Christmas Day.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 4.7,
                    count: 500
                    }
                }])

        render(
            <BrowserRouter>
                <Store.CartProvider>
                        <Cart />
                </Store.CartProvider>
            </BrowserRouter>
        )

        const subTotal = screen.queryByTestId('subtotal')

        expect(Number(subTotal.textContent.slice(1)).toFixed(2)).toBe("188.24")
    })
    it("correctly calculating tax", () => {

        Store.useCart = jest.fn(() => [{
                    id: 1,
                    title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
                    price: 109.95,
                    description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 3.9,
                    count: 120
                    }
                },
                {
                    id: 2,
                    title: "Mens Casual Premium Slim Fit T-Shirts ",
                    price: 22.3,
                    description: "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 4.1,
                    count: 259
                    }
                },
                {
                    id: 3,
                    title: "Mens Cotton Jacket",
                    price: 55.99,
                    description: "great outerwear jackets for Spring/Autumn/Winter, suitable for many occasions, such as working, hiking, camping, mountain/rock climbing, cycling, traveling or other outdoors. Good gift choice for you or your family member. A warm hearted love to Father, husband or son in this thanksgiving or Christmas Day.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 4.7,
                    count: 500
                    }
                }])

        render(
            <BrowserRouter>
                <Store.CartProvider>
                        <Cart />
                </Store.CartProvider>
            </BrowserRouter>
        )

        const tax = screen.queryByTestId('tax')

        expect(Number(tax.textContent.slice(1)).toFixed(2)).toBe("9.41")
    })

    it("correctly calculating total", () => {

        Store.useCart = jest.fn(() => [{
                    id: 1,
                    title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
                    price: 109.95,
                    description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 3.9,
                    count: 120
                    }
                },
                {
                    id: 2,
                    title: "Mens Casual Premium Slim Fit T-Shirts ",
                    price: 22.3,
                    description: "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 4.1,
                    count: 259
                    }
                },
                {
                    id: 3,
                    title: "Mens Cotton Jacket",
                    price: 55.99,
                    description: "great outerwear jackets for Spring/Autumn/Winter, suitable for many occasions, such as working, hiking, camping, mountain/rock climbing, cycling, traveling or other outdoors. Good gift choice for you or your family member. A warm hearted love to Father, husband or son in this thanksgiving or Christmas Day.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 4.7,
                    count: 500
                    }
                }])

        //const subTotal = Store.useCart().reduce((prevValue, currentValue) => {
            //return prevValue + currentValue.price
        //}, 0)

        const cart = Store.useCart()

        const cartSubtotal = cart.reduce((prevSubtotal, product) => prevSubtotal + product.price * product.quantity, 0)

        const cartTax = cartSubtotal * 0.05

        const cartTotal = cartSubtotal + cartTax

        render(
            <BrowserRouter>
                <Store.CartProvider>
                        <Cart />
                </Store.CartProvider>
            </BrowserRouter>
        )

        const total = screen.queryByTestId('total')

        expect(Number(total.textContent.slice(1)).toFixed(2)).toBe("197.65")
    })
    it("correctly calculating total when user increments or decrements product quantity", async () => {

        Store.useCart = jest.fn(() => [{
                    id: 1,
                    title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
                    price: 109.95,
                    description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 3.9,
                    count: 120
                    }
                },
                {
                    id: 2,
                    title: "Mens Casual Premium Slim Fit T-Shirts ",
                    price: 22.3,
                    description: "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 4.1,
                    count: 259
                    }
                },
                {
                    id: 3,
                    title: "Mens Cotton Jacket",
                    price: 55.99,
                    description: "great outerwear jackets for Spring/Autumn/Winter, suitable for many occasions, such as working, hiking, camping, mountain/rock climbing, cycling, traveling or other outdoors. Good gift choice for you or your family member. A warm hearted love to Father, husband or son in this thanksgiving or Christmas Day.",
                    category: "men's clothing",
                    image: "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg",
                    quantity: 1,
                    rating: {
                    rate: 4.7,
                    count: 500
                    }
                }])

        //const subTotal = Store.useCart().reduce((prevValue, currentValue) => {
            //return prevValue + currentValue.price
        //}, 0)

        const user = userEvent.setup()
        
        render(
            <BrowserRouter>
                <Store.CartProvider>
                        <Cart />
                </Store.CartProvider>
            </BrowserRouter>
        )

        const [firstIncrementProductQuantityButton]= screen.queryAllByText("+")

        await user.click(firstIncrementProductQuantityButton)

        const total = screen.queryByTestId('total')

        expect(Number(total.textContent.slice(1)).toFixed(2)).toBe("313.10")
    })


})