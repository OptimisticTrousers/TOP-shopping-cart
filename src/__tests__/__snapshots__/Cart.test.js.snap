// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem snapshot 1`] = `
<DocumentFragment>
  <div
    class="cart-container"
  >
    <div
      class="cart"
    >
      <h2
        class="cart-title"
      >
        Your Cart
      </h2>
      <div
        class="cost"
      >
        <p
          class="subtotal"
        >
          Subtotal
        </p>
        <p
          class="amount"
          data-testid="subtotal"
        >
          0
        </p>
      </div>
      <div
        class="cost"
      >
        <p
          class="delivery"
        >
          Delivery
        </p>
        <p
          class="amount"
        >
          $0.00
        </p>
      </div>
      <div
        class="cost"
      >
        <p
          class="tax"
        >
          Tax
        </p>
        <p
          class="amount"
        >
          + $14.00
        </p>
      </div>
      <div
        class="cost"
      >
        <p
          class="total"
        >
          Total
        </p>
        <p
          class="amount"
        >
          $78.13
        </p>
      </div>
    </div>
    <div
      class="buttons"
    >
      <button
        type="button"
      >
        Check Out
        <img
          alt="right arrow"
          class="right-arrow"
          src="right-arrow.svg"
        />
      </button>
      <a
        href="/shop"
      >
        <button
          type="button"
        >
          <img
            alt="left arrow"
            class="left-arrow"
            src="left-arrow.svg"
          />
           Back To Products
        </button>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
